src/analysis/souffle/taint.dl

```diff
@@ -171,4 +171,16 @@ predicateEval($CP_TagPresence(ap, owner, tag), 1) :-
 predicateEval($CP_TagPresence(ap, owner, tag), 0) :-
   checkSubExprs($CP_TagPresence(ap, owner, tag)), !mayHaveTag(ap, owner, tag).
 
+// This helper relation is used to make it very clear that predicateEval can be
+// only 1 xor 0 on a $CP_TagPresenceDynamicOwner.
+.decl tagPresentForSomeOwner(ap: AccessPath, tag: Tag)
+tagPresentForSomeOwner(ap, tag) :-
+  checkSubExprs($CP_TagPresenceDynamicOwner(ap, tag)),
+  ownsAccessPath(owner, ap), mayHaveTag(ap, owner, tag).
```

Comment:
This seems to completely disconnect the principal from the tag. We should actually verify that for every owner `owner` of an access path `ap`, `mayHaveTag(ap, owner, tag)` is true?  However, `tagPresentForSomeOwner` will be true if any owner has a tag?

Doesn't this weaken the existing checks?

Reply:
I don't think it does. See what we previously did at https://github.com/google-research/raksha/pull/269/files#r789881510 .

```C++
 constexpr absl::string_view kCheckHasTagFormat =
        R"(isCheck("$0", "$1"). check("$0", owner, "$1") :-
  ownsAccessPath(owner, "$1"), $2.)";
```

What that says is that we introduce a row in the `check` relation for any owner of the given `AccessPath` that also owns that tag. If the `check` row was present, we used that to populate `testFails` at this location https://github.com/google-research/raksha/pull/269/files#r789884147 as follows:

```C++
testFails(cat(check_index, "-", owner, "-", path)) :-
  isCheck(check_index, path), ownsAccessPath(owner, path),
  !check(check_index, owner, path).
```

That just says that the test fails if there does not exist an `owner` of the `path` for which `check` is populated.

Looking at my implementation, I populate `tagPresentForSomeOwner` if there exists an `owner` of the `AccessPath` indicated in the `CP_TagPresenceDynamicOwner` that also owns the `tag` we expect to be present. This is then used to populate `predicateEval` for that `CP_TagPresenceDynamicOwner` as true or false.

I think the two notations are actually semantically identical. If I'm missing an aspect that I've weakened, though, happy to change this to strengthen the enforcement of `CP_TagPresenceDynamicOwner`s.

Reply:
```
testFails(cat(check_index, "-", owner, "-", path)) :- 
  isCheck(check_index, path), ownsAccessPath(owner, path), 
  !check(check_index, owner, path).
```
> That just says that the test fails if there does not exist an owner of the path for which check is populated.

Just to make sure we are on the same page, the check is not simply populated. It is a conditional relation as follows: 
``` 
check(index, owner, path) :- 
  ownsAccessPath(owner, path), mayHaveTag(path, owner, "some-tag-from-particle-spec").
```

Therefore, the test will fail even if _any_ owner of the path disagrees that it has the corresponding tag. However, in the new version, the test will succeed even if some owner of the path disagrees that it has the corresponding tag. no?

Reply:
Good point. I have updated the code to make it insist that all owners agree that the tag is present and the number of owners is greater than 1.